### PR TITLE
Filter console messages before adding them to internally

### PR DESCRIPTION
**Pull request by randellhodges**
Add the ability to apply one or more regular expressions to messages to filter out those we do not want to see. It would be nice to have some way to tell klipper/moonraker not to send these to us in the first place, however, different subscribers may want different filters applied and that could get ugly. 

If that ability is added in the api in the future, this can be removed or tweaked to tell the api what to deliver.

https://github.com/meteyou/mainsail/issues/51

image

Simple temperature filter:
`^(?:ok\s+)?(B|C|T\d*):`

I'm still green with vue and vuetify. Inside ConsolePanel.vue, if I did not use blur, it would update $store with every keypress. Since the expressions are cached when evaluated, I didn't want a race between messages coming in and compiling partial regular expressions.

Thinking about it, probably need a way to validate the filters are valid regular expressions, or in the filterMessage method, handle those that don't compile? This may need to be resolved before merging. It works how it is, but only the happy path.

### CODE REVIEW

**Comment by randellhodges** [comment]
I added some rudimentary validation. I am not sure how to tie the validation together with the button in interface.vue. Presently, it won't prevent the save, but it won't save invalid filters at least.

Also changed to not calculate the expressions on demand, but calculate them when the incoming state data has filters specified and is being stored. Not sure if this is better or not.

**Comment by randellhodges** [comment]
@meteyou Any feedback?

File: src/components/panels/Settings/ConsolePanel.vue [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    ></v-textarea>
+                </v-col>
+            </v-row>
+        </v-card-text>
+    </v-card>
+</template>
+
+<script>
+
+    function getFiltersFromText(text) {
```
Comment (meteyou) [inline src/components/panels/Settings/ConsolePanel.vue:34]: you can move this function into methods

File: src/components/panels/Settings/ConsolePanel.vue [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        let inputValue = text.trim();
+        let filters = [...new Set(inputValue.split('\n'))];
+        return filters;
+    }
+
+    function getFilterErrors(filters) {
```
Comment (meteyou) [inline src/components/panels/Settings/ConsolePanel.vue:40]: same as getFiltersFromText. you can move this function into methods

File: src/store/mutations.js [hunk 3]
```diff
@@ -4,6 +4,29 @@ import { findDirectory } from '../plugins/helpers';
 import { colorArray, colorChamber, colorHeaterBed/*, temperaturChartSampleLength*/ } from './variables';
 import defaultPrinter from './printer';
 
+let consoleFilterCache = [];
+
+function updateConsoleFilterCache(filters) {
```
Comment (meteyou) [inline src/store/mutations.js:9]: you can move this function to plugins/helpers and only import these into the mutations

File: src/store/mutations.js [hunk 4]
```diff
@@ -4,6 +4,29 @@ import { findDirectory } from '../plugins/helpers';
...
+                const expr = new RegExp(filter);
+                consoleFilterCache.push((t) => expr.test(t));
+            } catch {
+                console.error('Error compiling filter: ' + filter);
+            }
+        }
+    });
+}
+
+function filterConsoleMessage(text) {
```
Comment (meteyou) [inline src/store/mutations.js:23]: you can move this function to plugins/helpers and only import these into the mutations

**Comment by randellhodges** [comment]
Thank you for in depth feedback. I'll get these resolved as soon as I can.

**Comment by LazaroFilm** [comment]
What about just a "Show temperatures in the console" toggle button? I doubt we need a full RegEx filter available to the user. That would remove the need to check if the expression is good, simplify the code and streamline the user experience (they wouldn't need to copy/paste that RegEx themselves.

**Comment by meteyou** [comment]
pls can you update your PR? if you need help with the new .env.development.example feel free to ping me. i will write a doc file in the next days.

**Comment by jperrin72** [comment]
Hey, this is a great addition to mainsail, any chab-nce this could be merge ? thx

**Comment by meteyou** [comment]
i added it by my own... https://github.com/meteyou/mainsail/commit/3c217e1161e093df90f8467067e932a428d44f2a

**Pull request closed without merging**